workers.py

```diff
@@ -1,49 +1,103 @@
+from contextlib import suppress
 import logging
+import os
 from threading import Thread, current_thread
-from json import dumps
+from tempfile import NamedTemporaryFile
+from uuid import uuid4
 
 import requests
 
-import s3
+logger = logging.getLogger()
+
+
+try:
+    # Try to import a provided parser. This can be AI service specific.
+    from parser import parse
+except ImportError:
+    logger.warning('No parser provided, fallback used!')
+
+    def parse(_):
+        """Fallback parser.
+
+        Used when no data preprocessor is provided.
+        """
+        return {}
 
 logger = logging.getLogger()
-HEADERS = {'Content-type': 'application/json'}
+CHUNK = 10240
+MAX_RETRIES = 3
 
 
-def download_and_pass_data_thread(filesystem, bucket, uri, next_service):
+def download_job(source_url: str, source_id: str, dest_url: str) -> None:
     """Spawn a thread worker for data downloading task.
 
     Requests the data to be downloaded and pass it to the next service
+    :param source_url: Data source location
+    :param source_id: Data identifier
+    :param dest_url: Location where the collected data should be received
     """
-    def worker():
+    # When source_id is missing, create our own
+    source_id = source_id or str(uuid4())
+
+    def worker() -> None:
+        """Download, extract data and forward the content."""
         thread = current_thread()
-        logger.info('%s: Worker started', thread.name)
+        logger.debug('%s: Worker started', thread.name)
+
+        # Fetch data
         try:
-            # Fetch data
-            s3_data = s3.fetch(filesystem, bucket, uri)
-            logger.info(
-                '%s: Downloaded records %s',
-                thread.name,
-                s3_data.shape
-            )
+            resp = requests.get(source_url, stream=True)
```

Comment:
How would this GET work without any authentication to the Data source?

Reply:
Why do you think it wouldn't work? Please look up messages on Kafka and upload service docs.

Reply:
As far as I know, the credentials are provided within the url. At least the messages which comes from the upload service do that:

```json
{
    "id": null,
    "url": "https://insights-upload-perm-test.s3.amazonaws.com/8c0c420de3ab4a92b832b0f1031e5908?AWSAccessKeyId=<REDACTED>&Signature=<REDACTED>&Expires=1544701539 ",
    "service": null,
    "payload_id": "8c0c420de3ab4a92b832b0f1031e5908",
    "account": null,
    "principal": null,
    "b64_identity": null,
    "rh_account": null,
    "rh_principal": null
}
```

I've redacted the `AWSAccessKeyId` and `Signature` from the message, but the format is the same. It's designed for easy access.

Reply:
Yes, understood what you mean. Thanks.

But what if the download fails for whatever reason? 
I guess this is a similar case that we discussed in the publisher PR - failure to Upload/implement Retry
Something like that would be expected here as well, right?

Reply:
I agree, a retry should be added. I've implemented the retry logic for the POST later on, but not here. Oh, boy. I'll fix that. Thanks!

Reply:
Feel free to add the Retry for GET in a follow-up PR